### PR TITLE
perf(workspace): ship the JavaScript without tsc's layout

### PR DESCRIPTION
**Pull request by ofri-peretz**
## Summary

- Every emitted `.js` is re-written through esbuild's `minifyWhitespace` in `scripts/build-package.ts`. Indentation alone was ~32% of a compiled rule file.
- **Ecosystem: 3283 kB → 2063 kB of shipped JavaScript (−37%).** A package install drops ~28% on disk.
- Deliberately **not** minification — identifiers keep their names, string contents are untouched, the syntax tree is not rewritten.

### Honest framing of the win

| metric | before → after |
|---|---|
| emitted JS (disk) | 3283 → 2063 kB (**−37%**) |
| package install / packagephobia | **−28%** |
| gzipped tarball (download) | **−8%** |

Gzip was already compressing the repeated indentation, so the download barely moves. Disk size is what this ecosystem competes on, so it's still worth doing — but I'm not going to claim a bandwidth win that isn't there.

### Why not full minification

Measured at another **4 kB gzipped**, and it would cost two things worth more: rule `meta` (messages, schema, docs URLs) is what the docs site and `--print-config` read, and a stack trace from inside a rule should still name its function.

## Verification

Built output vs the **published** `eslint-plugin-secure-coding@3.4.3`:

- identical lint findings, including `messageId`
- identical rule names (28)
- **0** differences across every rule's `meta`, `messages`, `schema`
- `configs.recommended` identical

Plus: full build 33/33, all 58 turbo test tasks, 1070 script tests, artifact gate clean.

## Ordering

Runs **after** the rule-barrel transform — that one pattern-matches tsc's emitted shape and wouldn't match once the layout is gone.

## Regression locks

**New:** `check-published-artifacts.ts` fails closed on indented JS. The naive "any indented line" test false-positives on multi-line template literals in rule messages (string contents are never stripped), so it requires both a high absolute count and a high proportion. Verified by re-indenting a package's dist: exit 1 with the regression, exit 0 clean.

**Fixed:** this change silently blinded the existing lazy-rules lock. Two of its three assertions matched tsc's text shape (`^const x_1 = require("./rules/y");$`); stripping the layout made them permanently unmatchable, so the lock would have kept passing while verifying nothing. Rewritten to assert the invariant that survives formatting — how many rule modules load on require (0 for lazy plugins, >0 for `eslint-plugin-jwt` where public re-exports force eager loading). Confirmed by mutation.

## Dependencies

`esbuild` is declared as a root devDependency rather than left resolving transitively through vitest — a phantom dependency is exactly what `verify:runtime-deps` exists to catch.

## Deliberately not included

Two larger size levers were evaluated and **rejected as functionality risks**:

- **`scslre` → optional peer** (−1448 kB on secure-coding): measured to lose **4 of 5 ReDoS detections**. Only the nested-quantifier case survives on heuristics. A security rule quietly protecting less while reporting clean is the wrong trade — and `eslint-plugin-security`, the package we'd beat on size, has no NFA analysis at all.
- **`oxc-resolver` → optional peer** (−1628 kB on import-next): only 2 of 55 rules need it, but both are in `recommended`, so it breaks the plugin out of the box unless those rules leave the preset. That's a product decision, not a build one.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/build-package.ts [hunk 1]
```diff
@@ -318,6 +319,65 @@ if (isPluginPackage) {
...
+      console.log(
+        `build-package(${pkg.name}): stripped JS whitespace — ${(before / 1024).toFixed(0)} kB -> ${(after / 1024).toFixed(0)} kB (-${saved}%).`,
+      );
+    }
+  } catch (error) {
+    // Same posture as the comment-strip pass: a degraded optimisation, not a
+    // broken build. Ship the readable JS rather than block every local build.
+    console.error(
+      `build-package(${pkg.name}): whitespace strip failed (${(error as Error).message}); shipping indented JS.`,
+    );
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply scripts/build-package.ts:387]: Correct, fixed in 205e9fbd.

The pass now transforms every file into memory and writes only once all of them have succeeded. Transforming and writing in one walk looked equivalent and was not — a throw on file N left 1..N-1 already written stripped, so `dist` ended up mixed while the catch reported "shipping indented JS" and the gate flagged only the remainder. Buffering costs ~2 MB for the largest package and makes the pass all-or-nothing, so the message is now true.

File: scripts/check-published-artifacts.ts [hunk 2]
```diff
@@ -88,6 +88,37 @@ type Violation = {
...
+const INDENT_MIN_LINES = 20;
+const INDENT_MIN_RATIO = 0.25;
+const INDENTED_LINE = {
+  test(source: string): boolean {
+    const lines = source.split('\n');
+    if (lines.length < INDENT_MIN_LINES) return false;
+    const indented = lines.filter((l) => /^ {2,}\S/.test(l)).length;
+    return (
+      indented >= INDENT_MIN_LINES &&
+      indented / lines.length >= INDENT_MIN_RATIO
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply scripts/check-published-artifacts.ts]: Right, and sharper than I'd appreciated — I'd calibrated that ratio against *unstripped* files without thinking about what stripping does to the denominator. Fixed in 205e9fbd by dropping the heuristic entirely.

The gate now tests **idempotency**: it re-runs the exact transform the build applies and compares sizes. A stripped file is a fixed point; an unstripped one collapses. Your edge case disappears because the template literal is byte-identical on both sides and cancels out — string contents are never touched.

Measured margin: 0.0% change re-stripping an already-stripped file, 27-41% on an indented one. Threshold sits at 5%.

Worth recording what this found. The new check immediately flagged six packages as unstripped after a full build, which turned out not to be a gate bug: `scripts/build-package.ts` was in no task's `inputs` and not a `globalDependency`, so editing the program that writes every published artifact changed not one cache key. Turbo restored `dist` built by the old script and said nothing. That hole silently covered the comment-strip pass, the `.d.ts` prune and the lazy rule-barrel transform too. Now a global dependency, locked in `turbo-cache-inputs-lock.test.ts`, verified by `0 cached, 35 total` after touching the script.
